OpenSim/Common/Path.cpp

```diff
@@ -0,0 +1,223 @@
+    }
+    _path.insert(_path.begin() + pos, pathElement);
+}
+
+void Path::erasePathElement(size_t pos)
+{
+    if (pos > getPathLength() - 1) {
+        OPENSIM_THROW(Exception, "erasePathElement: pos is out of range of elements");
+    }
+
+    _path.erase(_path.begin() + pos);
+}
+
+std::vector<std::string> Path::formAbsolutePathVec(Path* otherPath) 
+{
+    if (_isAbsolute) {
+        return _path;
+    }
+    
+    if (!otherPath->_isAbsolute) {
+        OPENSIM_THROW(Exception, "getOatherPathVec: otherPath must be an absolute path");
+    }
+
+    std::vector<std::string> pathVec;
+    for (auto pathElement : otherPath->_path) {
+        pathVec.push_back(pathElement);
+    }
+
+    for (auto pathElement : _path) {
+        pathVec.push_back(pathElement);
+    }
+
+    return pathVec;
+}
+
+std::vector<std::string> Path::formRelativePathVec(Path* otherPath) 
+{
+    if (!this->_isAbsolute || !otherPath->_isAbsolute) {
+        OPENSIM_THROW(Exception, "getRelativePathVec: both paths must be absolute to find relative path");
+    }
+
+    size_t thisPathLength = this->getPathLength();
+    size_t otherPathLength = otherPath->getPathLength();
+
+    // find how many elements at the head are in common
+    size_t searchLength = std::min(thisPathLength, otherPathLength);
+    bool match = true;
+    size_t ind = 0;
+    while (match && ind < searchLength) {
+        if (this->_path[ind] != otherPath->_path[ind]) {
+            match = false;
+        }
+        else {
+            ++ind;
+        }
+    }
+
+    // ind is the first index that doesn't match. first add any necessary
+    // "..", then add the remainder of the pathElements to get to otherPath
+    std::vector<std::string> pathVec;
+    
+    size_t numDotDot = otherPathLength - ind;
+    for (size_t i = 0; i < numDotDot; ++i) {
+        pathVec.push_back("..");
+    }
+    for (; ind < thisPathLength; ++ind) {
+        pathVec.push_back(_path[ind]);
+    }
+
+    return pathVec;
+}
+
+
+void Path::cleanPath() 
+{
+    size_t numPathElements = getPathLength();
+    size_t i = 0;
+    while (i < numPathElements) {
+        // remove any "." element
+        if (_path[i] == ".") {
+            erasePathElement(i);
+            --numPathElements;
+        }
+
+        // if it's not the first element, and the previous element is not ".."
+        // remove the ".." element and the previous element
+        else if (_path[i] == "..") {
+            if (i != 0 && _path[i-1] != "..") {
+                erasePathElement(i);
+                erasePathElement(i - 1);
+                numPathElements -= 2;
+                --i;
+            }
+            else {
+                ++i;
+            }
+        }
+
+        // otherwise, move on
+        else {
+            ++i;
+        }
+    }
+
+    if (!_path.empty() && _path[0] == ".." && _isAbsolute) {
+        OPENSIM_THROW(Exception, "absolute path cannot start with ''..''");
```

Comment:
Most of OpenSim's exception messages use proper-ish sentences. Consider starting exception messages with a capital letter.